### PR TITLE
feat(ext): support non-canonical HTTP/1 reason phrases

### PR DESCRIPTION
**Pull request by acfoltzer**
Add a new extension type `hyper::ext::ReasonPhrase` gated by a new feature flag `http1_reason_phrase`. When enabled, store any non-canonical reason phrases in this extension when parsing responses, and write this reason phrase instead of the canonical reason phrase when emitting responses.

Reason phrases are a disused corner of the spec that implementations ought to treat as opaque blobs of bytes. Unfortunately, real-world traffic sometimes does depend on being able to inspect and manipulate them.

My main outstanding question is: should this new feature flag be included in `full`? Excluding it means this patch also has to modify the CI configuration in order to run the `client` and `server` tests.

### CODE REVIEW

File: src/ext/h1_reason_phrase.rs
Comment (acfoltzer) [inline src/ext/h1_reason_phrase.rs:37]: :bike::derelict_house: Not particularly tied to this name, but tried to match the already-existing `ext` types which plainly state their purpose without mentioning their associated protocol version.

File: src/ffi/http_types.rs
Comment (acfoltzer) [inline src/ffi/http_types.rs:28]: I figured we may as well reuse the public type for the FFI bits.

File: src/proto/h1/role.rs
Comment (acfoltzer) [inline src/proto/h1/role.rs:366]: Here's where it starts getting a bit messy. To reduce the overall `cfg`ness of this code, I'm relying on `const` evaluation to ensure this does not have an effect on performance when the flag is not enabled.

**Review by acfoltzer — COMMENTED** [review]
So the tricky question is: is this worth a feature flag? Here are the performance numbers for the `cargo bench` suite (red is with the new feature disabled, green is with it enabled): https://gist.github.com/acfoltzer/c64e92ac95c390cb1c08624ecf61bfdf

It looks to me like roughly a wash, though there's a fair amount of noise in the sample. The problem is that none of the existing benches use the custom reason phrase, so it's hard to see the overall effect that existing users of `"full"` might get if they're already receiving responses with non-canonical phrases.

From an API cleanliness perspective I would prefer rolling this extension into the existing `"http1"` flag, but I also don't want to be introducing undue surprises for users of `client` who are not currently having to bear the cost of either the comparison against the canonical reason phrase or the allocation to store a custom one when sighted.

File: Cargo.toml
Comment (seanmonstar) [inline Cargo.toml:90]: Yea, so let's talk about if we need a feature. My initial gut feeling is that it seems like such a small feature that it doesn't pull its weight, just include it. Especially if the benchmarks don't show a noticeable difference.

File: src/ext/h1_reason_phrase.rs
Comment (seanmonstar) [reply src/ext/h1_reason_phrase.rs:37]: Yea the name seems fine to me. I don't think `Http1ReasonPhrase` is better 👀 .

File: src/ext/h1_reason_phrase.rs
Comment (seanmonstar) [inline src/ext/h1_reason_phrase.rs:56]: I've seen `from_static` in a lot of places, but never a `from_static_str`. Usually I'd expect people could either stick a `b` prefix on their string, or could call `as_bytes()`.

File: src/ext/h1_reason_phrase.rs
Comment (seanmonstar) [inline src/ext/h1_reason_phrase.rs:46]: I'd leave this off for now, just getting the bytes and being able to use `std::str::from_utf8()` is what I'd expect most people to do.

File: src/ext/h1_reason_phrase.rs
Comment (seanmonstar) [inline src/ext/h1_reason_phrase.rs:51]: Looking at this and the other constructors, I think we probably need to verify there's no illegal bytes in here that would mess up the HTTP/1 message.

**Review by seanmonstar — CHANGES_REQUESTED** [review]
Very good start, I appreciate the thoroughness of the docs and tests 🤩

File: src/ext/h1_reason_phrase.rs
Comment (acfoltzer) [reply src/ext/h1_reason_phrase.rs:51]: Ack, that's a really good catch. I'll convert some of the existing `From` impls to `TryFrom`, and expose an `unsafe` constructor for unchecked creation.

Comment (seanmonstar) [reply src/ext/h1_reason_phrase.rs:51]: Is the unchecked one needed yet? A `from_static` can likely just panic, similar to `HeaderName::from_static`. The verification can happen at compile time for most static strings.

File: Cargo.toml
Comment (acfoltzer) [reply Cargo.toml:90]: I'd be in favor of including it. I think it'd take some very unusual traffic (like maxing out reason phrase length on every response) to cause a noticeable difference. It also helps that the cost is incurred on responses rather than on requests which are usually more concerning as a DoS vector.

I'll go ahead and fold it into `http1` but will keep it in a separate commit while this is under review.

File: src/ext/h1_reason_phrase.rs
Comment (acfoltzer) [reply src/ext/h1_reason_phrase.rs:51]: We need to check when constructing at runtime, because the server role will happily splat out whatever bytes are there into the status line. If someone put non-spec characters into the byte string this would lead to Problems.

The unchecked one is there for the client role to save the validation step; it already knows the characters are valid since it made it through parsing.

Comment (acfoltzer) [reply src/ext/h1_reason_phrase.rs]: Implemented these interfaces in 699c4417, added tests in 4ca1eee9

File: src/ext/h1_reason_phrase.rs
Comment (acfoltzer) [reply src/ext/h1_reason_phrase.rs]: Removed in 699c4417

File: src/ext/h1_reason_phrase.rs
Comment (acfoltzer) [reply src/ext/h1_reason_phrase.rs]: Removed in 699c4417

File: Cargo.toml
Comment (acfoltzer) [reply Cargo.toml]: Made this change in d621c741, which led to a pleasant reduction in the amount of awkwardly `cfg`ed code :sparkles:

**Comment by seanmonstar** [comment]
@acfoltzer was this waiting on anything else, or just another review from me?

**Comment by acfoltzer** [comment]
@seanmonstar yep, though it looks like I'll need to fix some merge conflicts first. I'll try to get to that soon

**Comment by acfoltzer** [comment]
@seanmonstar rebased and squashed, incorporating the proposed change to fold this into the `http1` feature rather than introducing an entirely new feature just for the reason phrases. I think it's ready for a final review.

**Review by seanmonstar — APPROVED** [review]

**Pull request merged**
